stores/settings.go

```diff
@@ -0,0 +1,281 @@
+package stores
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"go.sia.tech/renterd/api"
+	sql "go.sia.tech/renterd/stores/sql"
+	"go.uber.org/zap"
+)
+
+const (
+	SettingGouging = "gouging"
+	SettingPinned  = "pinned"
+	SettingS3      = "s3"
+	SettingUpload  = "upload"
+)
+
+func (s *SQLStore) GougingSettings(ctx context.Context) (gs api.GougingSettings, err error) {
+	err = s.fetchSetting(ctx, SettingGouging, &gs)
+	return
+}
+
+func (s *SQLStore) UpdateGougingSettings(ctx context.Context, gs api.GougingSettings) error {
+	return s.updateSetting(ctx, SettingGouging, gs)
+}
+
+func (s *SQLStore) PinnedSettings(ctx context.Context) (ps api.PinnedSettings, err error) {
+	err = s.fetchSetting(ctx, SettingPinned, &ps)
+	return
+}
+
+func (s *SQLStore) UpdatePinnedSettings(ctx context.Context, ps api.PinnedSettings) error {
+	return s.updateSetting(ctx, SettingPinned, ps)
+}
+
+func (s *SQLStore) UploadSettings(ctx context.Context) (us api.UploadSettings, err error) {
+	err = s.fetchSetting(ctx, SettingUpload, &us)
+	return
+}
+
+func (s *SQLStore) UpdateUploadSettings(ctx context.Context, us api.UploadSettings) error {
+	return s.updateSetting(ctx, SettingUpload, us)
+}
+
+func (s *SQLStore) S3Settings(ctx context.Context) (ss api.S3Settings, err error) {
+	err = s.fetchSetting(ctx, SettingS3, &ss)
+	return
+}
+
+func (s *SQLStore) UpdateS3Settings(ctx context.Context, ss api.S3Settings) error {
+	return s.updateSetting(ctx, SettingS3, ss)
+}
+
+// MigrateV2Settings migrates the settings from the old format to the new,
+// migrating the existing settings over to the new types and removing the old
+// settings. If a setting is not present in the database it will be set to its
+// default setting. If an existing setting is not valid, the default will be
+// used and a warning will get logged.
+func (s *SQLStore) MigrateV2Settings(ctx context.Context) error {
```

Comment:
This is not consistent with how we do migrations. I'd write a regular database migration for this in raw SQL to make sure we only run it once and don't add "hidden" migrations.